crates/ethrpc/src/buffered.rs

```diff
@@ -84,23 +85,50 @@ where
         tokio::task::spawn(batched_for_each(config, calls, move |batch| {
             let inner = inner.clone();
             async move {
-                let (mut requests, mut senders): (Vec<_>, Vec<_>) = batch
-                    .into_iter()
-                    .filter(|(_, _, sender)| !sender.is_canceled())
-                    .map(|(id, request, sender)| ((id, request), sender))
-                    .unzip();
+                let (mut requests, mut trace_ids, mut senders): (Vec<_>, Vec<_>, Vec<_>) =
+                    itertools::multiunzip(
+                        batch
+                            .into_iter()
+                            .filter(|(_, _, _, sender)| !sender.is_canceled())
+                            .map(|(id, request, trace_id, sender)| {
+                                ((id, request), trace_id, sender)
+                            }),
+                    );
                 match requests.len() {
                     0 => (),
                     1 => {
-                        let ((id, request), sender) = (requests.remove(0), senders.remove(0));
-                        let result = inner.send(id, request).await;
+                        let ((id, request), trace_id, sender) =
+                            (requests.remove(0), trace_ids.remove(0), senders.remove(0));
+                        let result = match trace_id {
+                            Some(trace_id) => {
+                                observe::request_id::set_task_local_storage(
+                                    trace_id,
+                                    inner.send(id, request),
+                                )
+                                .await
+                            }
+                            None => inner.send(id, request).await,
```

Comment:
is it possible the scenario in which doesn't have a `trace-id`?

Reply:
Because the task's storage is optional, all the trace IDs are optional.